flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -355,6 +370,178 @@ static bool allSymbolsInExprPresentInMap(const Fortran::lower::SomeExpr &expr,
   return true;
 }
 
+/// Generate a load of a value from an address. Beware that this will lose
+/// any dynamic type information for polymorphic entities.
+static fir::ExtendedValue genLoad(fir::FirOpBuilder &builder,
+                                  mlir::Location loc,
+                                  const fir::ExtendedValue &addr) {
+  return addr.match(
+      [](const fir::CharBoxValue &box) -> fir::ExtendedValue { return box; },
+      [&](const fir::UnboxedValue &v) -> fir::ExtendedValue {
+        if (fir::unwrapRefType(fir::getBase(v).getType())
+                .isa<fir::RecordType>())
+          return v;
+        return builder.create<fir::LoadOp>(loc, fir::getBase(v));
+      },
+      [&](const fir::MutableBoxValue &box) -> fir::ExtendedValue {
+        return genLoad(builder, loc,
+                       fir::factory::genMutableBoxRead(builder, loc, box));
+      },
+      [&](const fir::BoxValue &box) -> fir::ExtendedValue {
+        return genLoad(builder, loc,
+                       fir::factory::readBoxValue(builder, loc, box));
+      },
+      [&](const auto &) -> fir::ExtendedValue {
+        fir::emitFatalError(
+            loc, "attempting to load whole array or procedure address");
+      });
+}
+
+/// create the zero value of a given the numerical or logical \p type (`false`
+/// for logical types).
+static mlir::Value createZeroValue(fir::FirOpBuilder &builder,
+                                   mlir::Location loc, mlir::Type type) {
+  mlir::Type i1 = builder.getIntegerType(1);
+  mlir::Value zero;
+  if (type.isa<fir::LogicalType>() || type == i1) {
+    zero = builder.createBool(loc, false);
+    zero = builder.createConvert(loc, type, zero);
```

Comment:
```suggestion
    return builder.createConvert(loc, type, zero);
```

Can get rid of the threaded `zero` variable, the `else` parts of `else if`s, and the `else` entirely.